codegen/masm/src/masm/program.rs

```diff
@@ -41,43 +81,204 @@ impl Program {
-            exe.push_back(start);
-            modules.insert(exe);
+            body.push(Op::Exec(entrypoint));
         }
+        exe.push_back(start);
+        exe
+    }
 
-        let segments = program.segments().clone();
+    /// Generate an executable module which when run expects the raw data segment data to be
+    /// provided on the advice stack in the same order as initialization, and the hash of a
+    /// procedure in the set of modules of the program that you wish to invoke, followed by any
+    /// operands to be passed as arguments.
+    fn generate_dyn_main(&self, emit_test_harness: bool) -> Box<Module> {
+        let mut exe = Box::new(Module::new(LibraryNamespace::Exec.into(), ModuleKind::Executable));
+        let start_id = FunctionIdent {
+            module: Ident::with_empty_span(Symbol::intern(LibraryNamespace::EXEC_PATH)),
+            function: Ident::with_empty_span(Symbol::intern(ProcedureName::MAIN_PROC_NAME)),
+        };
+        let start_sig = Signature::new([], []);
+        let mut start = Box::new(Function::new(start_id, start_sig));
 
-        Self {
-            modules,
-            segments,
-            entrypoint,
+        {
+            let body = start.body_mut();
+            self.emit_data_segment_initialization(body);
+            if emit_test_harness {
+                self.emit_test_harness(body);
+            }
+            body.push(Op::DynExec);
         }
+
+        exe.push_back(start);
+        exe
+    }
+
+    fn emit_test_harness(&self, block: &mut Block) {
+        // Advice Stack: [dest_ptr, num_words, ...]
+        block.push(Op::AdvPush(2)); // => [num_words, dest_ptr] on operand stack
+        block.push(Op::Exec("std::mem::pipe_words_to_memory".parse().unwrap()));
+        // Drop the commitment
+        block.push(Op::Drop);
+        // If we know the stack pointer address, update it to the value of `'write_ptr`, but cast
+        // into the Rust address space (multiplying it by 16). So a word address of 1, is equal to
+        // a byte address of 16, because each field element holds 4 bytes, and there are 4 elements
+        // in a word.
+        //
+        // If we don't know the stack pointer, just drop the `'write_ptr` value
+        if let Some(sp) = self.stack_pointer {
+            block.push(Op::U32OverflowingMulImm(16));
+            block.push(Op::Assertz);
+            // Align the stack pointer to a word boundary
+            let elem_addr = (sp / 4) + (sp % 4 > 0) as u32;
+            let word_addr = (elem_addr / 4) + (elem_addr % 4 > 0) as u32;
+            block.push(Op::MemStoreImm(word_addr));
+        } else {
+            block.push(Op::Drop);
+        }
+    }
+
+    /// Emit the sequence of instructions necessary to consume rodata from the advice stack and
+    /// populate the global heap with the data segments of this program, verifying that the
+    /// commitments match.
+    fn emit_data_segment_initialization(&self, block: &mut Block) {
```

Comment:
I was naively thinking of writing a rodata bytes directly into memory with `MemStoreImm`. Could you please elaborate on why go through the advice stack? I suspect it's an obvious reason, but I fail to see it.

Reply:
This function actually applies for more than testing, it is also how we initialize global memory on program start. The reason for using the advice stack is because it is vastly more efficient via `pipe_double_words_to_memory`, and because large amounts of memory would require many instructions to encode, making the program size much larger. By providing it via the advice stack, the actual initialization routine is compact, and we can load arbitrary amounts of data without affecting the program size.

The downside is that we have to orchestrate getting this data into the advice stack somehow, and there is no automation for this currently. My hope is that we solve this with future improvements to the VM.